packages/react/src/components/Flex/Flex.tsx

```diff
@@ -0,0 +1,119 @@
+import * as PropTypes from 'prop-types'
+import * as React from 'react'
+import * as _ from 'lodash'
+
+import { UIComponent, commonPropTypes } from '../../lib'
+import { ReactProps } from '../../types'
+import FlexItem from './FlexItem'
+
+export type FlexGap = 'gap.small' | 'gap.medium' | 'gap.large'
+export type FlexPadding = 'padding.medium'
+
+export interface FlexProps {
+  /** Defines if container should be inline element. */
+  inline?: boolean
+
+  /** Sets vertical flow direction. */
+  column?: boolean
+
+  /** Allows overflow items to wrap on the next container's line. */
+  wrap?: boolean
+
+  /** Controls items alignment in horizontal direction. */
+  hAlign?: 'start' | 'center' | 'end' | 'stretch'
+
+  /** Controls items alignment in vertical direction. */
+  vAlign?: 'start' | 'center' | 'end' | 'stretch'
+
+  /** Defines strategy for distributing remaining space between items. */
+  space?: 'around' | 'between' | 'evenly'
+
+  /** Defines gap between each two adjacent child items. */
+  gap?: FlexGap
+
+  /** Defines container's padding. */
+  padding?: FlexPadding
+
+  /** Enables debug mode. */
+  debug?: boolean
+
+  /** Orders container to fill all parent's space available. */
+  fill?: boolean
+}
+
+/**
+ * Arrange group of items aligned towards common direction.
+ */
+class Flex extends UIComponent<ReactProps<FlexProps>> {
+  static Item: typeof FlexItem
+
+  static Gap: any
+
+  static displayName = 'Flex'
+  static className = 'ui-flex'
+
+  static defaultProps = {
+    as: 'div',
+  }
+
+  public static propTypes = {
+    ...commonPropTypes.createCommon({
+      content: false,
+    }),
+
+    inline: PropTypes.bool,
+
+    column: PropTypes.bool,
+
+    wrap: PropTypes.bool,
+
+    hAlign: PropTypes.oneOf(['start', 'center', 'end', 'stretch']),
+    vAlign: PropTypes.oneOf(['start', 'center', 'end', 'stretch']),
+
+    space: PropTypes.oneOf(['around', 'between', 'evenly']),
+
+    gap: PropTypes.string,
+
+    padding: PropTypes.string,
+    fill: PropTypes.bool,
+
+    debug: PropTypes.bool,
+  }
+
+  renderChildren = gapClasses => {
+    const { column, gap, children } = this.props
+
+    let isFirst = true
+    return React.Children.map(children, (child: React.ReactElement<any>) => {
+      const childElement =
+        child.type && ((child.type as any) as typeof FlexItem).__isFlexItem
```

Comment:
This can be broken with `react-hot-loader`:
https://github.com/gaearon/react-hot-loader/tree/v4.6.5#checking-element-types
https://github.com/gaearon/react-hot-loader/blob/v4.6.5/src/utils.dev.js#L14

---

We can also try to use `react-is`:
```suggestion
        ReactIs.isElement(child) && ((child.type as any) as typeof FlexItem).__isFlexItem
```

Reply:
yes - and this is exactly why the flag-based approach is used, instead of just trying to compare types directly: as whoever will be the intermediate that would introduce proxied types (like `react-hot-loader`) - irregardless of this fact it is safe to assume that all the type's props will return the same values as original type (`react-hot-loader`, surely, adheres to that). 

Thus, in this aspect it should be pretty robust approach.